pkg/converter/convert_unix.go

```diff
@@ -1069,6 +1094,148 @@ func convertManifest(ctx context.Context, cs content.Store, oldDesc ocispec.Desc
+	return func(ctx context.Context, cs content.Store, _ ocispec.Descriptor, newDesc *ocispec.Descriptor) (*ocispec.Descriptor, error) {
+		desc := newDesc
+		if !images.IsManifestType(desc.MediaType) {
+			return desc, nil
+		}
+		if IsNydusBootstrap(*desc) {
+			return desc, nil
+		}
+		var err error
+		var labels map[string]string
+		switch desc.MediaType {
+		case ocispec.MediaTypeImageManifest, images.MediaTypeDockerSchema2Manifest:
+			var manifest ocispec.Manifest
+			labels, err = readJSON(ctx, cs, &manifest, *desc)
+			if err != nil {
+				return nil, errors.Wrap(err, "read manifest")
+			}
+
+			desc, err = writeJSON(ctx, cs, manifest, *desc, labels)
+			if err != nil {
+				return nil, errors.Wrap(err, "write manifest")
+			}
+			return desc, nil
+
+		case ocispec.MediaTypeImageIndex, images.MediaTypeDockerSchema2ManifestList:
+			var index ocispec.Index
+			labels, err = readJSON(ctx, cs, &index, *desc)
+			if err != nil {
+				return nil, errors.Wrap(err, "read manifest index")
+			}
+			for idx, maniDesc := range index.Manifests {
+				var manifest ocispec.Manifest
+				labels, err = readJSON(ctx, cs, &manifest, maniDesc)
+				if err != nil {
+					return nil, errors.Wrap(err, "read manifest")
+				}
+
+				newManiDesc, err := writeJSON(ctx, cs, manifest, maniDesc, labels)
+				if err != nil {
+					return nil, errors.Wrap(err, "write manifest")
+				}
+				index.Manifests[idx] = *newManiDesc
+			}
+			desc, err = writeJSON(ctx, cs, index, *desc, labels)
+			if err != nil {
+				return nil, errors.Wrap(err, "write manifest index")
+			}
+
+			return desc, nil
+
+		default:
+			return nil, errors.Errorf("unsupported media type %s", desc.MediaType)
+		}
+	}
+}
+
+func LayerReconvertFunc(opt UnpackOption) containerdReconverter.ConvertFunc {
+	return func(ctx context.Context, cs content.Store, desc ocispec.Descriptor) (*ocispec.Descriptor, error) {
+		if !images.IsLayerType(desc.MediaType) {
+			return nil, nil
+		}
+
+		// Skip the nydus bootstrap layer.
+		if IsNydusBootstrap(desc) {
+			logrus.Debugf("skip nydus bootstrap layer %s", desc.Digest.String())
+			return &desc, nil
+		}
+
+		ra, err := cs.ReaderAt(ctx, desc)
+		if err != nil {
+			return nil, errors.Wrap(err, "get reader")
+		}
+		defer ra.Close()
+
+		ref := fmt.Sprintf("convert-oci-from-%s", desc.Digest)
+		cw, err := content.OpenWriter(ctx, cs, content.WithRef(ref))
+		if err != nil {
+			return nil, errors.Wrap(err, "open blob writer")
+		}
+
+		var gw io.WriteCloser
+		var mediaType string
+		switch opt.Compressor {
+		case "gzip":
+			gw = gzip.NewWriter(cw)
+			mediaType = ocispec.MediaTypeImageLayerGzip
+		default:
+			gw, err = zstd.NewWriter(cw)
+			if err != nil {
+				return nil, errors.Wrap(err, "create zstd writer")
+			}
+			mediaType = ocispec.MediaTypeImageLayerZstd
+		}
+		var data bytes.Buffer
+		writer := io.Writer(&data)
+		uncompressedDgster := digest.SHA256.Digester()
+
+		err = Unpack(ctx, ra, writer, opt)
```

Comment:
It seems writing a large OCI tar blob into `data` (in memory), this will consume a lot of memory.